content/en/blog/_posts/2022-11-11-retroactive-default-storage-class.md

```diff
@@ -0,0 +1,163 @@
+
+With this new feature enabled we wanted to maintain this behavior but also be able to update the StorageClass name.
+With these constraints in mind, the feature changes the semantics of <code>nil</code>. If a default StorageClass is present, <code>nil<code> would translate to "Give me a default" and <code>""</code> would mean "Give me PersistentVolume that also has <code>""</code> StorageClass name." In the absence of a StorageClass, the behavior would remain unchanged.
+
+Summarizing the above, we've changed the semantics of <code>nil<code> so that its behavior depends on the presence or absence of a definition of default StorageClass.
+
+The tables below show all these cases to better describe when PVC binds and when it's StorageClass gets updated.
+
+<table>
+  <thead>
+     <tr>
+        <th colspan="2"></th>
+        <th>PVC storageClassName == <code>""</code></th>
+        <th>PVC storageClassName == <code>nil</code></th>
+     </tr>
+  </thead>
+  <tbody>
+     <tr>
+        <td rowspan="2">Without default class</td>
+        <td>PV storageClassName == <code>""</code></td>
+        <td>binds</td>
+        <td>binds</td>
+     </tr>
+     <tr>
+        <td>PV without storageClassName</td>
+        <td>binds</td>
+        <td>binds</td>
+     </tr>
+     <tr>
+        <td rowspan="2">With default class</td>
+        <td>PV storageClassName == <code>""</code></td>
+        <td>binds</td>
+        <td>class updates</td>
+     </tr>
+     <tr>
+        <td>PV without storageClassName</td>
+        <td>binds</td>
+        <td>class updates</td>
+     </tr>
+  </tbody>
+</table>
+
+## How to use it
+
+If you want to test the feature whilst it's alpha, you need to enable the relevant feature gate in the kube-controller-manager and the kube-apiserver. Use the <code>--feature-gates</code> command line argument:
+
+```bash
+--feature-gates="...,RetroactiveDefaultStorageClass=true"
+```
+
+### Test drive
+
+If you would like to see the feature in action and verify it works fine in your cluster here's what you can try:
+
+1. Define a basic PersistentVolumeClaim:
+```yaml
+apiVersion: v1
+kind: PersistentVolumeClaim
+metadata:
+  name: pvc-1
+spec:
+  accessModes:
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 1Gi
+```
+
+2. Create the PersistentVolumeClaim when there is no default StorageClass. The PVC won't provision or bind (unless there is an existing, suitable PV already present) and will remain in <code>Pending</code> state.
+```
+$ kc get pvc
+NAME      STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
+csi-pvc   Pending   
+```
+
+3. Configure one StorageClass as default.
+```
+$ kc patch sc/csi-hostpath-sc -p '{"metadata":{"annotations":{"storageclass.kubernetes.io/is-default-class":"true"}}}'
+storageclass.storage.k8s.io/csi-hostpath-sc patched
+```
```

Comment:
Whenever we suggest `hostPath` as an example, we *need* to point out that it's a terrible idea to make that storage driver available unrestricted in production clusters.

Either:
- make up a fictional CSI driver and use that instead
- clearly add text to warn readers

Reply:
Good point. Just changing the SC name to something generic like `my-storageclass` should be fine and we can avoid mentioning `hostPath` anywhere.